src/eventBus.ts

```diff
@@ -0,0 +1,137 @@
+    // ["hl_attr_define", id, rgb_attr, cterm_attr, info]
+    | IRedrawEventArg<
+          "hl_attr_define",
+          [
+              number,
+              {
+                  foreground?: number;
+                  background?: number;
+                  special?: number;
+                  reverse?: boolean;
+                  italic?: boolean;
+                  bold?: boolean;
+                  strikethrough?: boolean;
+                  underline?: boolean;
+                  underdouble?: boolean;
+                  underdotted?: boolean;
+                  underdashed?: boolean;
+                  undercurl?: boolean;
+                  blend?: number;
+              },
+              never,
+              [{ kind: "ui" | "syntax" | "terminal"; ui_name: string; hi_name: string }],
+          ]
+      >
+    // ["msg_show", kind, content, replace_last]
+    | IRedrawEventArg<
+          "msg_show",
+          [
+              (
+                  | ""
+                  | "confirm"
+                  | "confirm_sub"
+                  | "emsg"
+                  | "echo"
+                  | "echomsg"
+                  | "echoerr"
+                  | "lua_error"
+                  | "rpc_error"
+                  | "return_prompt"
+                  | "quickfix"
+                  | "search_count"
+                  | "wmsg"
+              ),
+              [number, string][],
+              boolean,
+          ]
+      >
+    // ["msg_showcmd", content]
+    | IRedrawEventArg<"msg_showcmd", [[number, string][]]>
+    // ["msg_showmode", content]
+    | IRedrawEventArg<"msg_showmode", [[number, string][]]>
+    // ["msg_ruler", content]
+    | IRedrawEventArg<"msg_ruler", [[number, string][]]>
+    // ["mode_info_set", cursor_style_enabled, mode_info]
+    | IRedrawEventArg<"mode_info_set", [boolean, { name: string; cursor_shape: "block" | "horizontal" | "vertical" }[]]>
+    // ["msg_history_show", entries]
+    | IRedrawEventArg<"msg_history_show", [string, [number, string][]][][]>
+    // ["msg_clear"]
+    | IRedrawEventArg<"msg_clear", []>
+    // ["mode_change", mode, mode_idx]
+    | IRedrawEventArg<"mode_change", [string, number]>
+    // ["cmdline_show", content, pos, firstc, prompt, indent, level]
+    | IRedrawEventArg<"cmdline_show", [[object, string][], number, string, string, number, number]>
+    // ["cmdline_hide"]
+    | IRedrawEventArg<"cmdline_hide", []>
+    // ["mouse_on"]
+    | IRedrawEventArg<"mouse_on", []>
+    // ["mouse_off"]
+    | IRedrawEventArg<"mouse_off", []>
+    | IRedrawEventArg<"wildmenu_show", [string[]]>
+)[];
+
+type EventsMapping = {
+    redraw: RedrawEventArgs;
+};
+
+export interface Event<T extends keyof EventsMapping = keyof EventsMapping> {
+    name: T;
+    data: EventsMapping[T];
+}
+
+export type EventBusData<T extends keyof EventsMapping> = EventsMapping[T];
+
+class EventBus implements Disposable {
+    private readonly emitter = new EventEmitter<Event>();
```

Comment:
Is the idea that all Nvim events are dispatched by this single `EventEmitter` ?

Reply:
Yeah I believe so. This will also let you write lua handlers for nvim events with the new api RFC